src/iosrv/ult.c

```diff
@@ -363,12 +363,35 @@ sched_ult2xs(int ult_type, int tgt_id)
 	switch (ult_type) {
 	case DSS_ULT_IOFW:
 	case DSS_ULT_MISC:
-		return (DSS_MAIN_XS_ID(tgt_id) + 1) % DSS_XS_NR_TOTAL;
+		if (!dss_helper_pool)
+			return (DSS_MAIN_XS_ID(tgt_id) + 1) % DSS_XS_NR_TOTAL;
+
+		if (dss_tgt_offload_xs_nr >= dss_tgt_nr)
+			return (dss_sys_xs_nr + dss_tgt_nr + tgt_id);
+		if (dss_tgt_offload_xs_nr > 0)
+			return (dss_sys_xs_nr + dss_tgt_nr +
+				tgt_id % dss_tgt_offload_xs_nr);
+		else
+			return ((DSS_MAIN_XS_ID(tgt_id) + 1) % dss_tgt_nr +
+				dss_sys_xs_nr);
 	case DSS_ULT_EC:
 	case DSS_ULT_CHECKSUM:
 	case DSS_ULT_COMPRESS:
 	case DSS_ULT_IO:
```

Comment:
it was added by one patch from @wangdi1 , the name of DSS_ULT_IO looks a kind of confuse to me also. 
Just checked its usage:  
1) seems some cases can reuse DSS_ULT_MISC
2) but the case of cont_epoch_discard_one() it access VOS (vos_discard) seems should use IO main XS
@wangdi1 ?

Reply:
cont_epoch_discard_one() is called by dss_thread_collective() which always create ults on VOS xstream (this ult2xs mapping function isn't called), however, I think we'd change this in case of dss_ult_create() is called with DSS_ULT_IO in the future.

Reply:
i see, agreed. just checked current all use cases of DSS_ULT_IO is collective.